mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java

```diff
@@ -19,21 +19,342 @@
+      Path resultPath = new Path(location, queryId);
+      FileSystem fs = Util.getFs(resultPath, conf);
+      RemoteIterator<LocatedFileStatus> fileStatuses = fs.listFiles(resultPath, false);
+      Set<String> filesToKeep = new HashSet<>();
+      Set<String> files = new HashSet<>();
+      while (fileStatuses.hasNext()) {
+        LocatedFileStatus status = fileStatuses.next();
+        String fileName = status.getPath().getName();
+        files.add(resultPath.toString() + "/" + fileName);
+        if (fileName.endsWith(COMMITTED_PREFIX)) {
+          LOG.debug("Reading committed file {}", fileName);
+          ClosedFileData cfd = readCommittedFile(table.io(), resultPath.toString() + "/" + fileName);
+          DataFiles.Builder builder = DataFiles.builder(PartitionSpec.unpartitioned())
+              .withPath(cfd.fileName)
+              .withFormat(cfd.fileFormat)
+              .withFileSizeInBytes(cfd.length)
+              .withMetrics(cfd.serializableMetrics.metrics());
+          append = append.appendFile(builder.build());
+          filesToKeep.add(cfd.fileName);
+        }
+      }
+      append.commit();
+      LOG.info("Iceberg write is committed for {} with files {}", table, filesToKeep);
+      files.removeAll(filesToKeep);
+      LOG.debug("Removing unused files: {}", files);
+      files.forEach(file -> table.io().deleteFile(file));
+      cleanupJob(jobContext);
+    }
+  }
+
+  private static void createCommittedFileFor(FileIO io, ClosedFileData closedFileData) throws IOException {
+    OutputFile commitFile = io.newOutputFile(closedFileData.fileName + COMMITTED_PREFIX);
+    ObjectOutputStream oos = new ObjectOutputStream(commitFile.createOrOverwrite());
+    oos.writeObject(closedFileData);
+    oos.close();
+    LOG.debug("Iceberg committed file is created {}", commitFile);
+  }
+
+  private static ClosedFileData readCommittedFile(FileIO io, String committedFileLocation) throws IOException {
+    try (ObjectInputStream ois = new ObjectInputStream(io.newInputFile(committedFileLocation).newStream())) {
+      return (ClosedFileData) ois.readObject();
+    } catch (ClassNotFoundException cnfe) {
+      throw new IOException("Can not parse committed file: " + committedFileLocation, cnfe);
+    }
+  }
+
+  private static final class ClosedFileData implements Serializable {
+    private String fileName;
+    private FileFormat fileFormat;
+    private Long length;
+    private SerializableMetrics serializableMetrics;
+
+    private ClosedFileData(String fileName, FileFormat fileFormat, Long length, Metrics metrics) {
+      this.fileName = fileName;
+      this.fileFormat = fileFormat;
+      this.length = length;
+      this.serializableMetrics = new SerializableMetrics(metrics);
+    }
+  }
+
+  /**
+   * We need this class, since Metrics in not Serializable (even though it implements the interface)
+   */
+  private static final class SerializableMetrics implements Serializable {
```

Comment:
I'm adding a metrics API For FS classes (readers, writers etc) to optionally implement: https://github.com/apache/hadoop/pull/2069

https://github.com/steveloughran/hadoop/blob/s3/HADOOP-16830-iostatistics/hadoop-common-project/hadoop-common/src/site/markdown/filesystem/iostatistics.md

While you can't use that (yet), it'd be good to design your stats so that they can be included

Reply:
Stats are inherited from Iceberg. For now I plan to avoid changing it.